Chisel/CHLFunctions.m

```diff
@@ -0,0 +1,65 @@
+//
+//  CHLFunctions.m
+//  Chisel
+//
+//  Copyright © 2016 Facebook. All rights reserved.
+//
+
+#import "CHLFunctions.h"
+
+// See vm_range_recorder_t
+static void enumerateRanges(task_t, void *, unsigned, vm_range_t *, unsigned);
+
+void CHLEnumerateAllocationsWithBlock(void (^block)(vm_range_t))
+{
+  /*
+   * === Overview ===
+   *
+   * To enumerate the set of live memory allocations ("ranges"), the pseudocode might look like:
+   *
+   *     for zone in all_zones()
+   *       for range in zone->ranges
+   *         callback(range)
+   *
+   * Unfortunately, the memory allocations within a zone are not exposed, so the above inner loop is not possible.
+   * Instead malloc zones provide a function that will loop over the allocations and call the given callback function.
+   */
+
+  /*
+   * === malloc API Notes ===
+   *
+   * Both `malloc_get_all_zones` and `vm_range_recorder_t` callback functions take a `task_t` and a `memory_reader_t`.
+   * However it turns out both of these can be NULL. This information isn't documented in malloc.h but can be seen in
+   * the libmalloc source.
+   *
+   * `memory_reader_t`:
+   * When NULL, the default memory reader is used which handles in process memory, and does not use the given `task_t`.
+   * `task_t`:
+   * Only required if used in a custom `memory_reader_t`, or used in a `vm_range_recorder_t` callback.
+   */
```

Comment:
I think we probably do want to keep these comments internal since they're about the malloc API that's being wrapped. This function provides a nice abstraction so that users don't have to worry about these details.